File: nth-prime/src/nth_prime.erl
-module(nth_prime).

-export([prime/1, is_prime/1]).

is_prime(2) ->
    true;
is_prime(N) when (N < 2) or (N rem 2 == 0) ->
    false;
is_prime(N) ->
    is_prime(N, 3).

is_prime(N, K) when N < K * K ->
    true;
is_prime(N, K) when N rem K == 0 ->
    false;
is_prime(N, K) ->
    is_prime(N, K + 2).

do_prime(N, N, Res) ->
    Res - 1;
do_prime(N, Ct, It) ->
    case is_prime(It) of
        true ->
            do_prime(N, Ct + 1, It + 1);
        false ->
            do_prime(N, Ct, It + 1)
    end.

prime(N) when N > 0 ->
    do_prime(N, 1, 3).

File: elixir/nth-prime/nth_prime.exs
defmodule Prime do
  @doc """
  Generates the nth prime.
  """
  @spec nth(non_neg_integer) :: non_neg_integer
  def nth(count) when count > 0 do
    Stream.iterate(2, &(&1 + 1))
    |> Enum.reduce_while(count, fn x, acc ->
      if is_prime(x) do
        acc = acc - 1

        if acc <= 0 do
          {:halt, x}
        else
          {:cont, acc}
        end
      else
        {:cont, acc}
      end
    end)
  end

  defp is_prime(n) do
    l = 2..n |> Enum.to_list() |> Enum.reverse() |> tl() |> Enum.reverse()
    [] == l || l |> Enum.all?(fn x -> rem(n, x) != 0 end)
  end
end
